### PR TITLE
implements copy!(dst::SparseVector, src) (fixes #30443)

### LINKED ISSUE #30443: cannot copy! SparseVectors
```
1.1.0-DEV.841> x=spzeros(1); copy!(x, spzeros(1));
ERROR: MethodError: no method matching empty!(::SparseVector{Float64,Int64})
Closest candidates are:
  empty!(::Base64.Buffer) at /home/ab/src/julia/build1.0/usr/share/julia/stdlib/v1.1/Base64/src/buffer.jl:15
  empty!(::BitSet) at bitset.jl:303
  empty!(::BitArray{1}) at bitarray.jl:1037
  ...
Stacktrace:
 [1] copy!(::SparseVector{Float64,Int64}, ::SparseVector{Float64,Int64}) at ./abstractarray.jl:655
 [2] top-level scope at none:0
```

it works for `SparseMatrixCSC` by the way. I don't know what is the "proper" way to fix this. I suspect that just implementing `empty!(x::SparseVector)` would lead to a suboptimal `copy!(dst::SparseVector, src::SparseVector)` with possibly redundant reallocations. So I'll leave this to someone more knowledgeable...

### PR DESCRIPTION
**Pull request by abraunst**
Not 100% sure if this is the way, but seems to me the most straightforward thing to do. Fixes #30443.

### CODE REVIEW

File: stdlib/SparseArrays/src/sparsevector.jl [hunk 1]
```diff
@@ -1935,6 +1935,16 @@ julia> dropzeros(A)
 """
 dropzeros(x::SparseVector; trim::Bool = true) = dropzeros!(copy(x), trim = trim)
 
+function copy!(dst::SparseVector, src::SparseVector)
+    dst.n == src.n || throw(ArgumentError("Vector should have the same length for copy!"))
```
Comment (rfourquet) [inline stdlib/SparseArrays/src/sparsevector.jl]: For `AbstractVector` in general, `dst` and `src` can be of different lengths, I don't see here the necessity to have this restriction. I think the two calls to `copy!` below will handle different lengths just fine.

File: stdlib/SparseArrays/src/sparsevector.jl [hunk 2]
```diff
@@ -1935,6 +1935,16 @@ julia> dropzeros(A)
...
+    copy!(dst.nzval, src.nzval)
+    copy!(dst.nzind, src.nzind)
+    return dst
+end
+
+function copy!(dst::SparseVector, src::AbstractVector)
+    copy!(dst, sparse(src))
```
Comment (rfourquet) [inline stdlib/SparseArrays/src/sparsevector.jl]: This seems too inefficient for a library function, as it creates an uncessary temporary allocation... 
I saw that in the issue you suspect implementing `empty!` for sparse vectors would be too inefficient: would you mind comparing the performance of that (using `empty!` and the default `copy!` implementation) versus your implementation of `copy!(dst::SparseVector, src::SparseVector)` above?

File: stdlib/SparseArrays/src/sparsevector.jl (hunk 1 above)
Comment (abraunst) [reply stdlib/SparseArrays/src/sparsevector.jl]: but `SparseVector` is immutable, so tere is no way of changing `n`. Should `SparseVector` be mutable?

File: stdlib/SparseArrays/src/sparsevector.jl (hunk 2 above)
Comment (abraunst) [reply stdlib/SparseArrays/src/sparsevector.jl]: I think that empty! cannot be really be implemented as long as `SparseVector` is immutable. Or am I misunderstanding?

File: stdlib/SparseArrays/src/sparsevector.jl (hunk 1 above)
Comment (rfourquet) [reply stdlib/SparseArrays/src/sparsevector.jl]: Yes sorry, I overlooked that! Then just update the error message to tell `"Sparse vectors should..."` to emphasize that it concerns only `SparceVector`.

File: stdlib/SparseArrays/src/sparsevector.jl (hunk 2 above)
Comment (rfourquet) [reply stdlib/SparseArrays/src/sparsevector.jl]: You are totally right, sorry again!

Comment (rfourquet) [reply stdlib/SparseArrays/src/sparsevector.jl]: Still, I'm hoping for a more efficient implementation here. That could be left for as a future optimization though.

Comment (abraunst) [reply stdlib/SparseArrays/src/sparsevector.jl]: I hear you. I just pushed a more efficient version by cannibalizing dense2sparsevec. I haven't thoughly tested or benchmarked it yet though. But let me know if you think it is more or less what you had in mind.

**Comment by abraunst** [comment]
It seems to be much more efficient than `.=` now which is encouraging...

```
1.2.0-DEV.71> x=sprand(10_000, 0.1); y=Vector(sprand(10_000,0.1));

1.2.0-DEV.71> x2=copy(x); @benchmark copy!($x2,$y)
BenchmarkTools.Trial: 
  memory estimate:  32 bytes
  allocs estimate:  1
  --------------
  minimum time:     10.056 μs (0.00% GC)
  median time:      10.476 μs (0.00% GC)
  mean time:        10.589 μs (0.00% GC)
  maximum time:     28.141 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.71> x2=copy(x); @benchmark $x2 .= $y
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     188.146 μs (0.00% GC)
  median time:      195.469 μs (0.00% GC)
  mean time:        197.989 μs (0.00% GC)
  maximum time:     603.007 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.71> x=sprand(10_000, 0.1); y=rand(10_000);

1.2.0-DEV.71> x2=copy(x); @benchmark copy!($x2,$y)
BenchmarkTools.Trial: 
  memory estimate:  32 bytes
  allocs estimate:  1
  --------------
  minimum time:     9.592 μs (0.00% GC)
  median time:      9.659 μs (0.00% GC)
  mean time:        9.820 μs (0.00% GC)
  maximum time:     32.602 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.71> x2=copy(x); @benchmark $x2 .= $y
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     315.776 μs (0.00% GC)
  median time:      327.449 μs (0.00% GC)
  mean time:        330.592 μs (0.00% GC)
  maximum time:     718.237 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.71> x2==y
true

1.2.0-DEV.71> x2=copy(x); @benchmark copy!($x2,$y)
BenchmarkTools.Trial: 
  memory estimate:  32 bytes
  allocs estimate:  1
  --------------
  minimum time:     9.606 μs (0.00% GC)
  median time:      9.716 μs (0.00% GC)
  mean time:        9.826 μs (0.00% GC)
  maximum time:     36.113 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.71> x2==y
true

1.2.0-DEV.71> x=sprand(10_000, 0.1); y=zeros(10_000);

1.2.0-DEV.71> x2=copy(x); @benchmark copy!($x2,$y)
BenchmarkTools.Trial: 
  memory estimate:  32 bytes
  allocs estimate:  1
  --------------
  minimum time:     5.391 μs (0.00% GC)
  median time:      5.394 μs (0.00% GC)
  mean time:        5.469 μs (0.00% GC)
  maximum time:     31.938 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     6

1.2.0-DEV.71> x2=copy(x); @benchmark $x2 .= $y
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     156.662 μs (0.00% GC)
  median time:      164.483 μs (0.00% GC)
  mean time:        166.832 μs (0.00% GC)
  maximum time:     870.523 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

```

**Comment by abraunst** [comment]
Also faster than `copy!(dst, sparse(src))` as it avoids temporaries:

```
1.2.0-DEV.71> x=sprand(10_000_000, 0.1); y=Vector(sprand(10_000_000,0.1));

1.2.0-DEV.71> x2=copy(x); @benchmark copy!($x2, $y)
BenchmarkTools.Trial: 
  memory estimate:  32 bytes
  allocs estimate:  1
  --------------
  minimum time:     15.300 ms (0.00% GC)
  median time:      15.498 ms (0.00% GC)
  mean time:        15.534 ms (0.00% GC)
  maximum time:     17.681 ms (0.00% GC)
  --------------
  samples:          322
  evals/sample:     1

1.2.0-DEV.71> x2=copy(x); @benchmark copy!($x2, sparse($y))
BenchmarkTools.Trial: 
  memory estimate:  19.07 MiB
  allocs estimate:  5
  --------------
  minimum time:     18.298 ms (0.00% GC)
  median time:      19.932 ms (1.12% GC)
  mean time:        19.798 ms (1.56% GC)
  maximum time:     22.407 ms (0.99% GC)
  --------------
  samples:          253
  evals/sample:     1
```

**Comment by mbauman** [comment]
Interesting — we already have `copyto!` defined for Sparse vectors, just not the new `copy!`.  The latter will resize vectors as appropriate — but we cannot resize sparse vectors.  So this looks good to me.

As an aside, it looks like we need to add a fast path to `copyto!` sparse arrays when it's a simple `X .= Y` in-place assignment — theoretically that really should be just as fast as calling `copyto!` directly.

File: stdlib/SparseArrays/src/sparsevector.jl [hunk 3]
```diff
@@ -378,13 +378,10 @@ sparsevec(a::AbstractSparseArray) = vec(a)
 sparsevec(a::AbstractSparseVector) = vec(a)
 sparse(a::AbstractVector) = sparsevec(a)
 
-function _dense2sparsevec(s::AbstractArray{Tv}, initcap::Ti) where {Tv,Ti}
+function _dense2indval!(nzind::Vector{Ti}, nzval::Vector{Tv}, s::AbstractArray{Tv}) where {Tv,Ti}
     @assert !has_offset_axes(s)
-    # pre-condition: initcap > 0; the initcap determines the index type
+    cap = length(nzind);
```
Comment (mbauman) [inline stdlib/SparseArrays/src/sparsevector.jl]: Let's also assert that `length(nzind) == length(nzval)` here as we assume that and use `@inbounds` with that assumption. It's an internal function so we needn't be too kind with the error message.

```suggestion
    cap = length(nzind);
    @assert cap == length(nzval)
```

File: stdlib/SparseArrays/src/sparsevector.jl [hunk 4]
```diff
@@ -403,7 +400,12 @@ function _dense2sparsevec(s::AbstractArray{Tv}, initcap::Ti) where {Tv,Ti}
         resize!(nzind, c)
         resize!(nzval, c)
     end
-    SparseVector(n, nzind, nzval)
+    nzind, nzval
```
Comment (mbauman) [inline stdlib/SparseArrays/src/sparsevector.jl]: I like explicit returns, especially when returning tuples.

```suggestion
    return (nzind, nzval)
```

**Comment by abraunst** [comment]
> Interesting — we already have `copyto!` defined for Sparse vectors, just not the new `copy!`. The latter will resize vectors as appropriate — but we cannot resize sparse vectors. So this looks good to me.
> 
> As an aside, it looks like we need to add a fast path to `copyto!` sparse arrays when it's a simple `X .= Y` in-place assignment — theoretically that really should be just as fast as calling `copyto!` directly.

Thank you @mbauman for the review. For `SparseVector` to `SparseVector` it seems that there is almost no gain, as  .= and copyto! seem almost equivalent on my setup (will try to benchmark better). But from `Vector` to `SparseVector`, the new copy! beats the two of them, by a large margin. So maybe it is worthwhile to implement `copyto!(dst::SparseVector, src::AbstractVector)` along the lines of this PR. But maybe it is better to do it in a separate PR, as it is purely an optimization.

```
1.2.0-DEV.75> x=sprand(10_000, 0.1); y=Vector(sprand(10_000,0.1));

1.2.0-DEV.75> x2=copy(x); @benchmark copyto!($x2, $y)                                                                             [3/1875]
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     191.975 μs (0.00% GC)
  median time:      200.059 μs (0.00% GC)
  mean time:        202.698 μs (0.00% GC)
  maximum time:     1.787 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.75> x2=copy(x); @benchmark $x2 .= $y
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     192.988 μs (0.00% GC)
  median time:      200.100 μs (0.00% GC)
  mean time:        202.814 μs (0.00% GC)
  maximum time:     638.732 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

1.2.0-DEV.75> x2=copy(x); @benchmark copy!($x2, $y)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     10.302 μs (0.00% GC)
  median time:      10.700 μs (0.00% GC)
  mean time:        10.822 μs (0.00% GC)
  maximum time:     29.741 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

**Review by mbauman — APPROVED** [review]

**Comment by abraunst** [comment]
> @abraunst this is awesome - but what about copying for sparse matrices? It's currently crazy slow:
> 
> ```julia
> using SparseArrays
> c = sprand(Bool,6000,18000, 0.01)
> d = view(c, 1:6000, rand(1:18000, 1800))
> copy(d);
> @time copy(d)
>   2.626470 seconds (12 allocations: 18.557 MiB, 0.42% gc time)
> ```
> 
> Do you have plans for this or should I open a new issue?

Is your issue about `view`s specifically or general for sparse matrices? I doubt that much can be done in this case (i.e. random access to a sparse matrix).

**Comment by mkborregaard** [comment]
My issue is on both but mostly views. They are generally ragged on the one axis and UnitRanges on the other. I know there are lots of issues on views into sparse matrices but it's one of my most common use cases - I already jump through lots of hoops to do simple operations (see https://github.com/EcoJulia/SpatialEcology.jl/blob/master/src/Sparse_matrixfunctions.jl ).
 2.6 seconds seems rather excessive for a view of this size...

**Comment by mbauman** [comment]
I say open a new issue with this specific case that you'd like to see optimized.

**Comment by mkborregaard** [comment]
Turned out I had 1.5 year old issue for this https://github.com/JuliaLang/julia/issues/21796

**Comment by mkborregaard** [comment]
So @abraunst @mbauman in this specific case, defining simply
```julia
mycopy(A::SubArray{R,S,T}) where R where S where T <: SparseMatrixCSC = d.parent[d.indices...]
```
lets you
```julia
using SparseArrays
c = sprand(Bool,6000,18000, 0.01);
d = view(c, 1:6000, rand(1:18000, 1800));
c1 = copy(d);
@time copy(d)
    2.626470 seconds (12 allocations: 18.557 MiB, 0.42% gc time)
c2 = mycopy(d);
@time mycopy(d)
    0.000564 seconds (12 allocations: 962.156 KiB)
c2 == c1
    true
```
Which is 3000 times faster (at least with `@time`). So it turns out it can indeed be optimised quite easily. Is this something to follow in terms of optimisation, or too naive?

**Comment by mkborregaard** [comment]
I'll move the above comment to the open issue

**Comment by mbauman** [comment]
Heck, make it a PR. :)

**Comment by mkborregaard** [comment]
Alright, on it :-)

**Pull request merged**